newIDE/app/src/Profile/SubscriptionDialog.js

```diff
@@ -33,319 +29,333 @@ import {
+          ),
+          rawError,
+          errorId: 'subscription-update-error',
+        });
+      } finally {
+        setIsChangingSubscription(false);
+      }
     } else {
-      showMessageBox(
-        i18n._(
-          t`Your subscription was properly cancelled. Sorry to see you go!`
-        )
-      );
-    }
-  };
+      // Changing the existing subscription.
+      const isSeamlessChange = canSeamlesslyChangeSubscription(subscription);
+      const confirmDialogTexts = isSeamlessChange
+        ? seamlesslyChangeConfirmationTexts
+        : cancelAndChangeConfirmationTexts;
+
+      const answer = await showConfirmation(confirmDialogTexts);
+      if (!answer) return;
+
+      if (isSeamlessChange) {
+        // Changing the existing subscription without asking for payment details again.
+        setIsChangingSubscription(true);
+        try {
+          await changeUserSubscription(getAuthorizationHeader, profile.id, {
+            planId: plan.planId,
+          });
+          await authenticatedUser.onRefreshUserProfile();
+          showMessageBox(
+            i18n._(
+              t`Congratulations, your new subscription is now active! You can now use the services unlocked with this plan.`
+            )
+          );
+        } catch (rawError) {
+          showErrorBox({
+            message: i18n._(
+              t`Your subscription could not be updated. Please try again later!`
+            ),
+            rawError,
+            errorId: 'subscription-update-error',
+          });
+        } finally {
+          setIsChangingSubscription(false);
+        }
+      } else {
+        // Changing the existing subscription by cancelling first.
+        setIsChangingSubscription(true);
+        try {
+          await changeUserSubscription(getAuthorizationHeader, profile.id, {
+            planId: '',
+          });
+          await authenticatedUser.onRefreshUserProfile();
+        } catch (rawError) {
+          showErrorBox({
+            message: i18n._(
+              t`Your subscription could not be cancelled. Please try again later!`
+            ),
+            rawError,
+            errorId: 'subscription-update-error',
+          });
+        } finally {
+          setIsChangingSubscription(false);
+        }
 
-  handleUpdatedSubscriptionFailure = (i18n: I18nType, rawError: Error) => {
-    this.setState({ isLoading: false });
-    showErrorBox({
-      message: i18n._(
-        t`Your subscription could not be updated. Please try again later!`
-      ),
-      rawError,
-      errorId: 'subscription-update-error',
-    });
+        // Then redirect as if a new subscription is being chosen.
+        setSubscriptionPendingDialogOpen(true);
+        Window.openExternalURL(
+          getRedirectToCheckoutUrl(planId, profile.id, profile.email)
+        );
+      }
+    }
   };
 
-  _renderPrice(plan: PlanDetails): React.Node {
+  const renderPrice = React.useCallback((plan: PlanDetails) => {
```

Comment:
hmm, a callback with no parameters, this can probably just be a function outside the React.FC

Reply:
Done